### PR TITLE
initial mtr-field commit

### PR DESCRIPTION
**Pull request by ozym**
A first go at bringing the device querying into the mtr.

The main jist of the application is an interface per device to do the metrics gathering, plus a wrapper of this to pass the results into a given function (which in this case allows sending them to the api).  Wrapping this is a main which has timers to request device metric queries, it gets all its device configuration from the api itself.

### CODE REVIEW

**Comment by gclitheroe** [comment]
Hi @crleblanc @junghao - a large PR from @ozym 

Please could you cast your eyes over it but not merge yet.  Any feedback to Mark.  I will have a look when I get a chance (maybe next week).

**Comment by junghao** [comment]
This is really cool.

Is it possible that we make each model under "field/" take its responsibility of init and register itself to main by creating init() and expanding the Prober interface?

Then we don't have to handle model level stuffs in main. eg:
https://github.com/GeoNet/mtr-field/blob/master/main.go#L28
https://github.com/GeoNet/mtr-field/blob/master/main.go#L159
https://github.com/GeoNet/mtr-field/blob/master/main.go#L235

Not sure if this can be done. (Probably there are some limitations from Golang?)

**Comment by ozym** [comment]
Yes, I've been iterating around this problem. 

One option I had thought of was to have a fixed set of options that are pretty universal for all models, i.e. a standard New() routine, and then have all other options handled via environment variables. The current version is a bit of both. I may try to see if I could make each model into its own package, and then New(), Init(), etc. all make a lot more sense for a specific model, perhaps under field/probe/xxxx/_probe_,go .... I may also see if I can group some similar models together and share/override some of the functions.

I'll do some experimenting, but I think I will still need the regexp somewhere in the main to select which `Probe` goes with each model -- but it could be simpler, or in an init() function.

**Comment by gclitheroe** [comment]
My overall comment is "I sort of get it but I think I would have a really hard time understanding it well enough to change it".  And to follow that up, is it a requirement - for others to be able to work on this?

There are 5 different ways that the application gets it's configuration:

* hard coded in main.go
* from command line flags which includes regular expressions
* from env var
* from the mtr api
* from a file of additional names or regular expressions.

This makes it extremely confusing as to how the program will work.  We should consider reducing the configuration complexity.  It may be useful to consider a "task service" from the mtr-api or some other source of truth.

Additionally their is code to handle SIGHUP which seems to just accelerate rereading of config which is happening on a timer anyway.

My impression is this is a macro service that is probably turned into a micro service via configuration to deploy it?  I had imagined something along a more micro service architecture.  For example a pinger program, deployed in a pinger container.  A voltage collector, possibly specific to a single device type, deployed in a container.  This should 

* help remove the need for so much configuration
* allow some stability on smaller services
* make understanding any one part of it somewhat easier.

I imagine there will be some needs for a library etc.  I find it most effective when these are pulled out as a maintenance task (pulling out the wrong abstraction to early is painful).

There are a lot of channels in use and it's pretty hard to follow.  I'm pretty confused by all the logic to stop the probes.  What's this needed for as in when does this have to be used?  Which is a little coupled to the logic to introduce jitter into when the probes happen.  

I'm not sure about the use of the channel to send the metrics to the mtr-api - what's the need for this?  It looks like sending the gathered metrics to the mtr-api is then rate limited internally by the chan length 1?    Why not just send in a go routine every time a probe gets a metric.  I can't see a reason to funnel them all back in or am I missing something?  It also looks like you adjust the timeouts and numbers of connections on the http transport so again I'm confused by the use of the rate limiting channel and the SIGTERM channel.

Associated to sending the metrics - in mtr.go `func (m MTR) Put(path string) error { ` You should consider in this send is some sort of back off and retry in the case of a server error (5xx).   

There is really severe deep `for` `if if else` and `switch` nesting in several of the probe functions.  I'm desperately confused here in many places.

There are several metrics being collected that we have no endpoints for.  Are these going to other systems?

There are publicly exported funcs in main.  It's not that important (you can't import main).

**Comment by ozym** [comment]
Working on a new workflow ..
